src/Autofac/Diagnostics/DiagnosticTracerBase.cs

```diff
@@ -0,0 +1,367 @@
+    /// <remarks>
+    /// <para>
+    /// Following events from a <see cref="System.Diagnostics.DiagnosticListener"/>
+    /// involves subscribing to that listener with an <see cref="IObserver{T}"/>
+    /// where the observable is a <see cref="KeyValuePair{K,V}"/> of <see cref="string"/>
+    /// event names and <see cref="object"/> event data.
+    /// </para>
+    /// <para>
+    /// This class helps with that by providing strongly typed event handlers
+    /// pre-wired to convert the event data into the right object type, allowing you
+    /// to focus on the handling logic and not the event/data parsing.
+    /// </para>
+    /// <para>
+    /// While technically you could subscribe to non-Autofac events using this
+    /// observer, it's not a general purpose mechanism - it's very much tailored
+    /// to Autofac.
+    /// </para>
+    /// </remarks>
+    /// <seealso cref="DefaultDiagnosticTracer"/>
+    public abstract class DiagnosticTracerBase : IObserver<KeyValuePair<string, object>>
+    {
+        /// <summary>
+        /// The list of event names to which this observer is subscribed.
+        /// </summary>
+        private readonly List<string> _subscriptions = new List<string>();
+
+        /// <summary>
+        /// Subscribes the observer to a particular named diagnostic event.
+        /// </summary>
+        /// <param name="diagnosticName">
+        /// The name of the event to which the observer should subscribe. Diagnostic
+        /// names are case-sensitive.
+        /// </param>
+        /// <remarks>
+        /// <para>
+        /// By default the observer is not subscribed to any events. You must
+        /// opt-in to any events you wish to handle. You may use
+        /// <see cref="EnableAll"/> to subscribe to all Autofac events at once.
+        /// </para>
+        /// </remarks>
+        /// <seealso cref="EnableAll"/>
+        /// <seealso cref="Disable"/>
+        public void Enable(string diagnosticName)
+        {
+            if (diagnosticName == null)
+            {
+                throw new ArgumentNullException(nameof(diagnosticName));
+            }
+
+            if (!_subscriptions.Contains(diagnosticName))
+            {
+                _subscriptions.Add(diagnosticName);
+            }
+        }
+
+        /// <summary>
+        /// Subscribes the observer to all Autofac events.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// By default the observer is not subscribed to any events. You must
+        /// opt-in to any events you wish to handle. This method helps you
+        /// to quickly subscribe to all the Autofac events at once. You may use
+        /// <see cref="Enable"/> to subscribe to individual events.
+        /// </para>
+        /// </remarks>
+        /// <seealso cref="Enable"/>
+        /// <seealso cref="Disable"/>
+        public void EnableAll()
+        {
+            _subscriptions.Add(DiagnosticEventKeys.MiddlewareStart);
```

Comment:
For a micro-perf gain while capturing diagnostics, I'd suggest re-ordering these events so they are ordered by 'most frequently invoked'.

So I'd put all the failures at the end, then put the request items before the operation ones.

Reply:
I reordered these by:
- Middleware events (most often - start, then success)
- Request events (second most often - start, then success)
- Operation events (third most often - start, then success)
- Failures by most to least frequent (middleware, request, operation)

Hopefully that helps a bit.